runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/ClueScrollWorldOverlay.java

```diff
@@ -0,0 +1,306 @@
+			{
+				LocalPoint localLocation = LocalPoint.fromWorld(client, crypticClue.getLocation());
+
+				if (localLocation != null)
+				{
+					markLocationWithImage(graphics, localLocation, SPADE_IMAGE);
+				}
+			}
+
+			// Mark NPC
+			if (crypticClue.getNpc() != null)
+			{
+				markNpc(graphics, crypticClue.getNpc());
+			}
+
+			// Mark game object
+			if (crypticClue.hasObject())
+			{
+				net.runelite.api.Point mousePosition = client.getMouseCanvasPosition();
+
+				GameObjectQuery query = new GameObjectQuery();
+				GameObject[] result = queryRunner.runQuery(query);
+
+				for (GameObject gameObject : result)
+				{
+					if (gameObject.getId() != crypticClue.getObjectId())
+					{
+						continue;
+					}
+
+					drawClickbox(graphics, gameObject.getClickbox(), mousePosition);
+					drawImageAtLocation(graphics, gameObject.getLocalLocation(), CLUE_SCROLL_IMAGE, IMAGE_Z_OFFSET);
+				}
+			}
+		}
+		else if (clue instanceof EmoteClue)
+		{
+			EmoteClue emoteClue = (EmoteClue) clue;
+
+			LocalPoint localLocation = LocalPoint.fromWorld(client, emoteClue.getLocation());
+
+			if (localLocation == null)
+			{
+				return null;
+			}
+
+			markLocationWithImage(graphics, localLocation, EMOTE_IMAGE);
+		}
+		else if (clue instanceof FairyRingClue)
+		{
+			FairyRingClue fairyRingClue = (FairyRingClue) clue;
+
+			LocalPoint localLocation = LocalPoint.fromWorld(client, fairyRingClue.getLocation());
+
+			if (localLocation == null)
+			{
+				return null;
+			}
+
+			markLocationWithImage(graphics, localLocation, SPADE_IMAGE);
+		}
+		else if (clue instanceof MapClue)
+		{
+			MapClue mapClue = (MapClue) clue;
+
+			LocalPoint localLocation = LocalPoint.fromWorld(client, mapClue.getLocation());
+
+			if (localLocation == null)
+			{
+				return null;
+			}
+
+			// Mark game object
+			if (mapClue.isObjectClue())
+			{
+				Region region = client.getRegion();
+				Tile[][][] tiles = region.getTiles();
+
+				Tile tile = tiles[client.getPlane()][localLocation.getRegionX()][localLocation.getRegionY()];
+
+				net.runelite.api.Point mousePosition = client.getMouseCanvasPosition();
+
+				for (GameObject gameObject : tile.getGameObjects())
+				{
+					if (gameObject != null)
+					{
+						drawClickbox(graphics, gameObject.getClickbox(), mousePosition);
+						drawImageAtLocation(graphics, localLocation, CLUE_SCROLL_IMAGE, IMAGE_Z_OFFSET);
+					}
+				}
+			}
+			else
+			{
+				markLocationWithImage(graphics, localLocation, SPADE_IMAGE);
+			}
+		}
+		else if (clue instanceof WorldPoint)
+		{
+			LocalPoint localLocation = LocalPoint.fromWorld(client, (WorldPoint) clue);
+
+			if (localLocation == null)
+			{
+				return null;
+			}
+
+			markLocationWithImage(graphics, localLocation, SPADE_IMAGE);
+		}
+
+		return null;
+	}
+
+	private void markNpc(Graphics2D graphics, String npcName)
+	{
+		Query query = new NPCQuery().nameContains(npcName);
+		NPC[] result = queryRunner.runQuery(query);
```

Comment:
It would be better to do this on a timer in the plugin and then have the overlay simply render that, instead of doing whole npc scans per frame.